Source/WebCore/dom/ScriptExecutionContextNativePromiseDispatcher.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2008 Apple Inc. All Rights Reserved.
+ * Copyright (C) 2012 Google Inc. All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "config.h"
+#include "ScriptExecutionContextNativePromiseDispatcher.h"
+
+namespace WebCore {
+
+ScriptExecutionContextNativePromiseDispatcher::ScriptExecutionContextNativePromiseDispatcher(ScriptExecutionContext& context, TaskSource source)
+    : m_queue(Queue::create(context, source))
+{
+}
+
+ScriptExecutionContextNativePromiseDispatcher::~ScriptExecutionContextNativePromiseDispatcher()
+{
+}
+
+void ScriptExecutionContextNativePromiseDispatcher::registerCallback(WTF::NativePromiseCallback& callback)
+{
+    if (RefPtr context = dynamicDowncast<WorkerGlobalScope>(m_queue->context()))
+        context->registerNativePromiseCallback(callback);
+}
+
+ScriptExecutionContextNativePromiseDispatcher::Queue::Queue(ScriptExecutionContext& context, TaskSource source)
+    : m_context(context)
+    , m_identifier(context.identifier())
+    , m_source(source)
+    , m_threadId(context.isWorkerGlobalScope() ? Thread::current().uid() : 0)
```

Comment:
MainThread::id is 1 as using in WorkQueue & RunLoop; if you kept that default it would simplify the isCurrent() implementation.